event.c

```diff
@@ -1987,6 +2003,14 @@ event_base_loop(struct event_base *base, int flags)
 
 		event_queue_make_later_events_active(base);
 
+		/* Invoke prepare watchers before polling for events */
+		EVBASE_RELEASE_LOCK(base, th_base_lock);
```

Comment:
Another question: when my test was a standalone executable, it seemed there was no problem calling `event_base_gettimeofday_cached()` from within a watcher callback. But when I converted it into `regress` testcases, it broke: the lock was already held in the call, and it failed:

```
[err] /usr/local/google/home/mergeconflict/github/libevent/evthread.c:239: Assertion lock->count == 1 failed in evthread_debug_lock_mark_locked

Program received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
51      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt    
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffff66453fa in __GI_abort () at abort.c:89
#2  0x00007ffff69e49bf in event_exit (errcode=-559030611) at /usr/local/google/home/mergeconflict/github/libevent/log.c:96
#3  0x00007ffff69e4fb4 in event_errx (eval=-559030611, fmt=0x7ffff69e8774 "%s:%d: Assertion %s failed in %s")
    at /usr/local/google/home/mergeconflict/github/libevent/log.c:153
#4  0x00007ffff69dd5e4 in evthread_debug_lock_mark_locked (mode=0, lock=0x715560)
    at /usr/local/google/home/mergeconflict/github/libevent/evthread.c:239
#5  0x00007ffff69dcea4 in debug_lock_lock (mode=0, lock_=0x715560) at /usr/local/google/home/mergeconflict/github/libevent/evthread.c:261
#6  0x00007ffff69cd683 in event_base_gettimeofday_cached (base=0x7150e0, tv=0x7fffffffda30)
    at /usr/local/google/home/mergeconflict/github/libevent/event.c:445
```

Releasing the lock before invoking the watchers fixes the issue, and it seemed like the right thing to do since we also release it when invoking user callbacks for active events, but I want to double-check with you that this makes sense.

Reply:
I also wonder why it worked at all when my test was a standalone executable...

Reply:
> when my test was a standalone executable, it seemed there was no problem calling event_base_gettimeofday_cached() from within a watcher callback

libevent do not thread safe by default (since this is overhead), you can enable it with `evthread_use_pthreads()` this is the first part, see below for the second

> But when I converted it into regress testcases, it broke: the lock was already held in the call, and it failed:

Second part is that `regress` enables lock debugging (`evthread_enable_lock_debugging()`) and that is why this assert pops up (you can disable it by installing `EVENT_NO_DEBUG_LOCKS` environment variable) instead of dead locking the program (since relocking of the non-recursive mutex is most of the time this is deadlock, but strictly speaking it is "undefined behavior")

>Releasing the lock before invoking the watchers fixes the issue, and it seemed like the right thing to do since we also release it when invoking user callbacks for active events, but I want to double-check with you that this makes sense

Indeed, my mistake that I didn't noticed this during looking into this yesterday (actually I didn't look for the correctness that time, just "conceptually")
Any callback that is called from the event loop should not hold the base mutex.

Reply:
Thinking a bit about overhead... I'm wondering if we should gate this new feature behind a compilation flag, like:
```C
#ifdef EVENT__ENABLE_WATCHERS
...
#endif
```
so that you don't pay any performance cost (especially around locking/unlocking) if you're not going to use the feature. What do you think?